src/auth/auth_gitlab.py

```diff
@@ -0,0 +1,339 @@
+            self.dump_sessions_to_file()
+            return False
+
+        self.user_states[user]['visit'] = now
+
+        # check gitlab response ttl, also check for stale updating (ttl*2)
+        if self.gitlab_update is not None:
+            stale = (self.user_states[user]['updated'] + max(self.gitlab_update*2, 60)) < now
+            ttl_expired = (self.user_states[user]['updated'] + self.gitlab_update) < now
+            updating_now = self.user_states[user]['updating'] is True
+            if ttl_expired and (not updating_now or stale):
+                if self.gitlab_group_support:
+                    self.do_update_groups(user, access_token)
+                else:
+                    self.do_update_user(user, access_token)
+
+        return True
+
+    def get_groups(self, user, known_groups=None):
+        if self.user_states.get(user) is None:
+            return []
+        if self.user_states[user]['groups'] is None:
+            return []
+        return self.user_states[user]['groups']
+
+    def logout(self, user, request_handler):
+        request_handler.clear_cookie('token')
+
+    def clean_expired_sessions(self):
+        now = time.time()
+        if self.session_expire:
+            for userData in list(self.user_states.keys()):
+                if (self.user_states[userData]['visit'] + self.session_expire) < now:
+                    LOGGER.debug("User %s session expired and removed" % userData)
+                    del self.user_states[userData]
+
+    def dump_sessions_to_file(self):
+        if self.gitlab_dump:
+            dumpFile = open(self.gitlab_dump, "w")
+            dumpFile.write(escape.json_encode(self.user_states))
+            dumpFile.close()
+            LOGGER.debug("Dumped state to file %s" % self.gitlab_dump)
+
+    def do_update_user(self, user, access_token):
+        self.user_states[user]['updating'] = True
+        tornado.ioloop.IOLoop.current().spawn_callback(self.update_user_state, user, access_token)
+
+    def do_update_groups(self, user, access_token):
+        self.user_states[user]['updating'] = True
+        tornado.ioloop.IOLoop.current().spawn_callback(self.update_group_list, user, access_token)
+
+    @gen.coroutine
+    def update_group_list(self, user, access_token):
+        group_list = yield self.read_groups(access_token)
+        if group_list is None:
+            LOGGER.error("Failed to refresh groups for %s" % user)
+            self.user_states[user]['state'] = "error"
+        else:
+            LOGGER.info("Groups for %s refreshed: " % user + str(group_list))
+            self.user_states[user]['groups'] = group_list
+        now = time.time()
+        self.user_states[user]['updating'] = False
+        self.user_states[user]['updated'] = now
+        self.user_states[user]['visit'] = now
+        self.clean_expired_sessions()
+        self.dump_sessions_to_file()
+        return
+
+    @gen.coroutine
+    def update_user_state(self, user, access_token):
```

Comment:
What does it method supposed to do? What kind of data does it load?

Reply:
This is request of user information. Same as at login point. Called `update_user_state` because we are only interesting in `state` field